### PR TITLE
Add request interception hooks

### PR DESCRIPTION
**Pull request by pimterry**
While working on https://github.com/resin-io/resin-ui/issues/400, it became clear the SDK currently isn't a usable replacement for our api calls in the UI, because it can't use Angular's built-in `$http` client internally. We depend on all requests going through that `$http` client so we can intercept them all in various places, e.g. to support offline status monitoring.

This PR adds Angular-inspired request/requestError/response/responseError hooks to resin-request, so that they can be exposed on resin-sdk shortly. These should have the same semantics as Angular's, which seem as good a model as any.

### CODE REVIEW

**Comment by pimterry** [comment]
Build fails because the new version of `fetch-mock` (used via resin-fetch-mock) has quietly starting expecting a global promise when it's run in browsers, and Phantom doesn't have one. Node tests in CI pass fine though, and my browsers tests do locally. I'm going to ignore these browser tests here for now, and get that sorted properly in fetch-mock itself soon.

**Comment by emirotin** [comment]
Can we just globally inject something like https://www.npmjs.com/package/es6-promise to make it happy? It won't add the bluebird sugar methods so won't mask any errors related to that.

File: lib/request.coffee [hunk 1]
```diff
@@ -97,6 +103,32 @@ module.exports = getRequest = ({ token, debug = false, retries = 0, isBrowser =
...
+		interceptRequestOrError(Promise.reject(requestError))
+
+	interceptResponse = (response) ->
+		interceptResponseOrError(Promise.resolve(response))
+
+	interceptResponseError = (responseError) ->
+		interceptResponseOrError(Promise.reject(responseError))
+
+	interceptRequestOrError = (initialPromise) ->
+		Promise.resolve(
```
Comment (emirotin) [inline lib/request.coffee]: Why do we need this wrapper? The inner thing is already a promise

File: lib/request.coffee [hunk 2]
```diff
@@ -97,6 +103,32 @@ module.exports = getRequest = ({ token, debug = false, retries = 0, isBrowser =
...
+			exports.interceptors.reduce (promise, interceptor) ->
+				promise.then(interceptor.request, interceptor.requestError)
+			, initialPromise
+		)
+
+	interceptResponseOrError = (initialPromise) ->
+		Promise.resolve(
+			exports.interceptors.slice().reverse().reduce (promise, interceptor) ->
+				promise.then(interceptor.response, interceptor.responseError)
```
Comment (emirotin) [inline lib/request.coffee]: if both are null-ish will `p.then(null, null)` resolve with the original promise value?

File: lib/request.coffee [hunk 3]
```diff
@@ -205,7 +244,57 @@ module.exports = getRequest = ({ token, debug = false, retries = 0, isBrowser =
...
+	# @public
+	#
+	# @description
+	# The current array of interceptors to use. Interceptors intercept requests made
+	# by calls to `.stream()` and `.send()` (some of which are made internally) and
+	# are executed in the order they appear in this array for requests, and in the
+	# reverse order for responses.
+	#
+	# @example
+	# resin.interceptors.push(
```
Comment (emirotin) [inline lib/request.coffee]: `resin` -> `resinRequest`?

File: tests/hooks.spec.coffee [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+
+			promise = request.send
+				url: 'https://original.com'
+			.get('body')
+
+			m.chai.expect(promise).to.eventually.become(requested: 'https://changed.com')
+
+		it 'should be able to asynchronously change a request before it is sent', ->
+			request.interceptors[0] = request: (request) ->
+				Promise.fromCallback (callback) ->
```
Comment (emirotin) [inline tests/hooks.spec.coffee]: NIT but can use http://bluebirdjs.com/docs/api/delay.html

File: tests/hooks.spec.coffee [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+			fetchMock.get 'https://500.com', (url, opts) ->
+				status: 500
+
+			request.interceptors[0] = responseError:
+				m.sinon.mock().throws(new Error('caught error'))
+
+			promise = request.stream
+				url: 'https://500.com'
+
+			m.chai.expect(promise).to.be.rejectedWith('caught error')
```
Comment (emirotin) [inline tests/hooks.spec.coffee]: Very nice test suite!

File: lib/request.coffee (hunk 2 above)
Comment (emirotin) [inline lib/request.coffee]: OK it works but shows a warning: https://monosnap.com/file/1uCMR5wTIKJtfoxQwm90SmJtKosfTE.png

So would change to

```
{ response, responseError } = interceptor 
return if (response or responseError) then promise.then(interceptor.response, interceptor.responseError) else promise`
```

(and same below)

File: lib/request.coffee (hunk 1 above)
Comment (pimterry) [inline lib/request.coffee]: I think this the result of reduce could be a non-bluebird promise, if a hook internally returns a native promise - this makes sure we definitely always get a bluebird promise at the end.

File: lib/request.coffee (hunk 2 above)
Comment (pimterry) [inline lib/request.coffee]: Done - destructuring here is a great idea.

File: lib/request.coffee (hunk 3 above)
Comment (pimterry) [inline lib/request.coffee]: Good spot. I've gone for `request` in the end, for consistency with the other examples.

File: tests/hooks.spec.coffee (hunk 4 above)
Comment (pimterry) [inline tests/hooks.spec.coffee]: Done

**Comment by pimterry** [comment]
All done - one comment on the `Promise.resolve` outstanding, let me know if that makes sense @emirotin.

I fixed the build by putting a patch in to `fetch-mock` directly, which seems to now all be working here very nicely, as I've just done a `resin-fetch-mock` release.

**Comment by emirotin** [comment]
Yeah the reasoning behind using `Promise.resolve` makes sense.

**Pull request merged**